rpc/grpc/server/src/connection.rs

```diff
@@ -160,354 +168,74 @@ impl Connection {
-                        match kaspa_rpc_core::NotifyPruningPointUtxoSetOverrideRequest::try_from(&notify_request) {
-                            Ok(request) => {
-                                let result = notifier
-                                    .clone()
-                                    .execute_subscribe_command(
-                                        listener_id,
-                                        Scope::PruningPointUtxoSetOverride(PruningPointUtxoSetOverrideScope::default()),
-                                        request.command,
-                                    )
-                                    .await;
-                                NotifyPruningPointUtxoSetOverrideResponseMessage::from(result).into()
-                            }
-                            Err(err) => NotifyPruningPointUtxoSetOverrideResponseMessage::from(err).into(),
-                        };
-                    KaspadResponse {
-                        id: 0,
-                        payload: Some(kaspad_response::Payload::StopNotifyingPruningPointUtxoSetOverrideResponse(response)),
-                    }
-                }
+    pub fn request_channel_size() -> usize {
+        256
+    }
 
-                _ => {
-                    return Err(GrpcServerError::InvalidSubscriptionPayload);
-                }
+    fn subscribe(&self, core_service: &DynRpcService, rpc_op: RpcApiOps) -> RequestSender {
+        let (sender, receiver) = mpsc_channel(Self::request_channel_size());
+        let handler = HandlerFactory::new_handler(rpc_op, self.clone(), core_service, self.inner.notifier.clone(), receiver);
+        handler.launch();
+        match self.inner.routing_map.write().insert(rpc_op, sender.clone()) {
+            Some(_) => {
+                // Overrides an existing route -- panic
+                error!("GRPC, Connection::subscribe overrides an existing value: {:?}, client: {}", rpc_op, self.identity());
+                panic!("GRPC, Tried to replace an existing route");
             }
-        } else {
-            return Err(GrpcServerError::InvalidSubscriptionPayload);
-        };
-        response.id = request.id;
+            None => {
+                trace!("GRPC, Connection::subscribe - {:?} route is registered, client:{:?}", rpc_op, self.identity());
+            }
+        }
+        sender
+    }
 
-        Ok(response)
+    async fn route_to_handler(&self, core_service: &DynRpcService, request: KaspadRequest) -> GrpcServerResult<()> {
+        // TODO: add appropriate error
+        if request.payload.is_none() {
+            debug!("GRPC, Route to handler got empty payload, client: {}", self);
+            return Err(GrpcServerError::InvalidRequestPayload);
+        }
+        let rpc_op = request.payload.as_ref().unwrap().into();
+        let sender = self.inner.routing_map.read().get(&rpc_op).cloned();
+        let sender = sender.unwrap_or_else(|| self.subscribe(core_service, rpc_op));
```

Comment:
There is a theoretic race-condition here if 2 threads try to add a new handler for the same op concurrently. `sender` will be `None` and they both will subscribe with the second one panicking. However this is not possible with current logic since a single task calls `route_to_handler` sequentially. Yet we should either document it or fix this by changing subscribe to use map `entry` api and adding the entry only if needed

Reply:
Let's opt for the entry API